File: cargo/crates/numbers-utils/src/operations/fibonacci_utils.rs
use num_bigint::BigUint;
use rand::Rng;

/// # `fibonacci`
///
/// This function calculates the nth Fibonacci number.
///
/// __Arguments:__
///
/// * `num: usize` - The nth Fibonacci number to calculate.
///
/// __Return:__
///
/// * `BigUint` - The nth Fibonacci number.
///
/// __Example:__
/// ```rust
/// use numbers_utils::operations::fibonacci_utils::fibonacci;
/// let fib = fibonacci(10);
/// println!("Fib[10] = {}", fib);
/// ```
///
/// __Output:__
/// ```txt
/// Fib[10] = 55
/// ```
pub fn fibonacci(num: usize) -> BigUint {
    let mut first_term = BigUint::from(0u8);
    let mut second_term = BigUint::from(1u8);

    for _ in 0..num {
        let temp = first_term.clone();
        first_term = second_term.clone();
        second_term += temp;
    }

    first_term
}

/// # `fibonacci_random`
///
/// This function generates a random number and returns its Fibonacci number.
///
/// __Return:__
///
/// * `String` - The Fibonacci number of the random number.
///
/// __Example:__
/// ```rust
/// use numbers_utils::operations::fibonacci_utils::fibonacci_random;
/// let fib = fibonacci_random();
/// println!("{}", fib);
/// ```
///
/// __Output:__
/// ```txt
/// Fib[10] = 55
/// ```
pub fn fibonacci_random() -> String {
    let mut rng = rand::thread_rng();
    let random_num: u32 = rng.gen_range(1..=50);

    let fib = fibonacci(random_num as usize);

    format!("Fib[{}] = {}", random_num, fib)
}

File: haskell/packages/numbers-utils/src/Numbers/Utils/Operations/FibonacciUtils.hs
module Numbers.Utils.Operations.FibonacciUtils (fibonacci, fibonacciRandom) where

import System.Random (randomRIO)

-- |
-- # @fibonacci@
--
-- The function @fibonacci@ is used to calculate the Fibonacci number.
--
-- __Arguments:__
--
-- * @num: Integer@ - The number to calculate the Fibonacci number.
--
-- __Return:__
--
-- * @Integer@ - The Fibonacci number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FibonacciUtils (fibonacci)
-- let fib = fibonacci 5
-- putStrLn $ "Fib[5] = " ++ show fib
-- @
--
-- __Output:__
--
-- @
-- "Fib[5] = 5"
-- @
fibonacci :: Integer -> Integer
fibonacci num = fibs !! fromIntegral num
    where
        fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

-- |
-- # @fibonacciRandom@
--
-- The function @fibonacciRandom@ is used to calculate the Fibonacci number of a
-- random number.
--
-- __Return:__
--
-- * @String@ - The Fibonacci number of the random number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FibonacciUtils (fibonacciRandom)
-- fib <- fibonacciRandom
-- putStrLn fib
-- @
--
-- __Output:__
--
-- @
-- "Fib[50] = 12586269025"
-- @
fibonacciRandom :: IO String
fibonacciRandom = do
    randomNum <- randomRIO (1, 50 :: Integer)
    let fib = fibonacci randomNum
    return $ "Fib[" ++ show randomNum ++ "] = " ++ show fib
